### PR TITLE
Add importer for 1Password's new export format (.1pux)

### PR DESCRIPTION
**Pull request by djsmith85**
## Type of change

- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective
With 1Password changing their export format from `.1pif` to `.1pux` in recent versions, this adds an importer to import 1pux into Bitwarden.

The 1pux format is essentially a zip archive including multiple files. The initial one of interest is the `export.data` file which contains the vault data.

Further information about the new 1pux format can be found here: https://support.1password.com/1pux-format/

**Rough overview of imported ciphers/fields:**
- Trashed items are ignored
- Empty fields are skipped

- **Login items**
  - Username
  - Password
  - TOTP
  - Password history
- **CreditCard items**
  - Cardholder
  - CC Number
  - CVV
  - Expiry date
  - Card brand is set using CC Number and only to our supported brands
- **Identity items**
  - Firstname
  - Initial
  - LastName
  - DefaultPhone
  - Company
  - Email
  - Username
  - Address
- Import of custom fields (loginfields and sections) for all CipherTypes
  - Fields marked as Date are saved as Unix Epoch and are converted to UTC string

- Favorites are set
- Import into folders based on the first entry in tags
- Import into collection based on the first entry in tags, if the import is ran from within an organization
- Folder and Collection names are captialized

- Category Mappings
1password supports various categories (Item types) which are saved in categoryUuid and are mapped as the following:
- Bitwarden Login Type:
  - Login
  - Database
  - Password 
  - WirelessRouter
  - Server
  - API Credentials

- Bitwarden CreditCard Type:
  - CreditCard
  - BankAccount

- Bitwarden Secure Note:
  - SecureNote
  - SoftwareLicense
  - EmailAccount
  - MedicalRecord

- Bitwarden Identity Type:
  - Identity
  - DriversLicense
  - OutdoorLicense
  - Membership
  - Passport
  - RewardsProgram
  - SocialSecurityNumber
  
1pux is prepared for including attachments but as of now, they are not included. This will be handled at a later point when 1password has added this feature.

Asana task: https://app.asana.com/0/1153292148278596/1201437354678085/f

## Code changes
- **common\src\importers\onepasswordImporters\types\onepassword1PuxImporterTypes.ts:** Added types definitions for the `export.data` file
- **common\src\importers\onepasswordImporters\onepassword1PuxImporter.ts:** The actual `export.data` importer for parsing and importing vaults.
- **\spec\common\importers\testData\onePassword1Pux\\*.ts:** Several files with test data using when executing the spec
- **\spec\common\importers\onepassword1PuxImporter.spec.ts:** Spec including several test cases to validate an import

## Testing requirements
Test scenarios will be listed on the web PR.

## Before you submit

- [X] I have checked for **linting** errors (`npm run lint`) (required)
- [X] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Comment by kspearrin** [comment]
Are we requiring users to unzip the file themselves before running our importer?

**Comment by djsmith85** [comment]
> Are we requiring users to unzip the file themselves before running our importer?

Not sure yet, wanna see if we are able to upload/unzip it in the web client and pass on the export.data to the importer. That would be the prefered solution.

**Comment by differsthecat** [comment]
Curious as to how this handles TOTP for free account imports; If we import to a free account, is that data saved and it is later unlocked for the user if they upgrade?

File: common/src/importers/onepasswordImporters/onepassword1PuxImporter.ts [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+  // email?: string | null;
+  // phone?: string | null;
+  // menu?: string | null;
+  // gender?: string | null;
+  // monthYear?: number | null;
+  // url?: string | null;
+  // address?: Address | null;
+  // creditCardType?: string | null;
+  // creditCardNumber?: string | null;
+  // reference?: string | null;
```
Comment (differsthecat) [inline common/src/importers/onepasswordImporters/onepassword1PuxImporter.ts:361]: These look like the fields for the `Value` interface; do we need them commented here?

File: common/src/importers/onepasswordImporters/onepassword1PuxImporter.ts [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+
+    if (this.isNullOrWhitespace(cipher.identity.company) && field.id === "company_name") {
+      cipher.identity.company = fieldValue;
+      return true;
+    }
+
+    // if (this.isNullOrWhitespace(cipher.identity.licenseNumber) && field.id === "membership_no") {
+    //   cipher.identity.licenseNumber = fieldValue;
+    //   return true;
+    // }
```
Comment (differsthecat) [inline common/src/importers/onepasswordImporters/onepassword1PuxImporter.ts:600]: Should this be uncommented or can we remove it?

File: common/src/importers/onepasswordImporters/onepassword1PuxImporter.ts (hunk 1 above)
Comment (djsmith85) [reply common/src/importers/onepasswordImporters/onepassword1PuxImporter.ts]: Removed. Just used that as a reference, so I didn't have to jump between files

**Review by differsthecat — APPROVED** [review]
Looks great, love the super thorough tests!

File: common/src/importers/onepasswordImporters/onepassword1PuxImporter.ts (hunk 2 above)
Comment (djsmith85) [reply common/src/importers/onepasswordImporters/onepassword1PuxImporter.ts]: I had the idea to map the membership-no into the licenseNumber field, but decided to keep it in the custom-fields along with the other unmapped ones.

**Comment by djsmith85** [comment]
> Looks great, love the super thorough tests!

Thank you, was loads of work, but this will hopefully aid, if users encounter any issues and encourage devs to add more tests 😉

**Pull request merged**